### PR TITLE
Add Validation Splits and Logging, Refactor Dataset Blueprints, and Improve Float8 Support

### PR DESCRIPTION
**Pull request by NSFW-API**
• Introduces separate train/val dataset groups in the blueprint config and uses them for both cache_latents.py and cache_text_encoder_outputs.py.  
• Extends hv_train_network.py with a validate() function that runs on the val_dataset_group each epoch, computes MSE loss, and logs val_loss via Accelerate (e.g., accelerator.log(...)).  
• Refactors config_utils.py to return separate train_dataset_group and val_dataset_group, combining them when needed (all_datasets).  
• Adds optional float8 fallback handling in token_refiner.py, safely casting float8 → float for calculations and then back to float8.  
• Adjusts cache skipping/keeping logic to handle old cache files, and changes the code to enumerate all_datasets instead of only train datasets.  
• Overall, this PR makes it possible to run a distinct validation pass each epoch, log validation performance, and unify caching for both train and val.

Screenshot 2025-01-26 093517

### CODE REVIEW

File: cache_text_encoder_outputs.py
Comment (NSFW-API) [inline cache_text_encoder_outputs.py:1]: Most of this file wasn't intended to change, `o1` just went a bit overboard. Will cut it down to only necessary changes.

Comment (kohya-ss) [reply cache_text_encoder_outputs.py:1]: Thank you, that's very helpful, it takes me some extra time to understand the changes.

Comment (NSFW-API) [reply cache_text_encoder_outputs.py:1]: @kohya-ss Should be ready for you to take a look.

File: hunyuan_model/token_refiner.py [hunk 1]
```diff
@@ -233,8 +233,18 @@ def forward(
         if mask is None:
             context_aware_representations = x.mean(dim=1)
         else:
-            mask_float = mask.float().unsqueeze(-1)  # [b, s1, 1]
```
Comment (NSFW-API) [inline hunyuan_model/token_refiner.py:236]: I was getting a crash on my 4090 about type casting I think, this resolved it.

**Comment by NSFW-API** [comment]
> Thank you! I think this is a great starting point.
> 
> As for the dataset settings, just an idea, how about adding the `is_val` (or `is_validation`) attribute to the dataset? This might help minimize modifications to cache_latents.py etc.

I missed this earlier. Will take a look at what it would involve but I don't see why that wouldn't work.

**Comment by NSFW-API** [comment]
I think I need to apply the same stable loss to the normal loss graph too, not just the validation loss.

**Comment by NSFW-API** [comment]
> Thank you! I think this is a great starting point.
> 
> As for the dataset settings, just an idea, how about adding the `is_val` (or `is_validation`) attribute to the dataset? This might help minimize modifications to cache_latents.py etc.

I've implemented this change, no longer needing to modify the caching scripts. Good call out!

**Comment by NSFW-API** [comment]
After re-testing, looks like the cache_latents.py script is failing. It might need some changes to accommodate the `is_val` change after all. Will resubmit once I make sure all parts are working as expected.

**Comment by NSFW-API** [comment]
Ok, just needed a small change to those caching files.

**Comment by Enyakk** [comment]
I am trying out the PR and it looks like great instrument to add to the toolkit. However two changes, if practical, would enhance the effectiveness:
- Large datasets might only have 1-2 epochs during the entirety of the training run so a steps_validate option would be appreciated
- Multiple named validation datasets in tensorboard: Training contents may have multiple goals. Being able to measure each goal against a defined validation dataset (ie. character A, concept B, regularization C) and display the current validation state would enhance productivity even further.

**Comment by NSFW-API** [comment]
I got reports that this isn't working when you don't provide a validation dataset, which it was previously as validation should of course be optional. Will see why that broke with the latest commit, also it seems I need to add support for JSON configs in addition to TOML. I missed that there are a variety of ways to specify datasets.

**Comment by niceguy4** [comment]
Any opportunities with revisiting this implementation? Validation seems to be an important element for verifying training. It appears to work well when used in conjunction with the tensor board loss chart.

**Comment by Sana-A-E** [comment]
What is the state of this? If I pull this branch, will the training work properly and log validation loss properly?
Are you guys still working on getting this merged into the main branch?
Honestly, having a validation loss chart would really help out in picking checkpoints. When training image generation models, you can just output a grid to compare the outputs of all the checkpoints, but you can't do the same with video, which makes comparing checkpoints a chore. Not to mention the time it would take to generate all those videos for comparison. 
With validation loss curve, instead of struggling to compare 20 versions of the checkpoint, it could be cut down to e.g. 4.
